agcaballero/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -57,106 +96,186 @@ boolean isReplaceableBy(Availability availability) {
-    // spots that match one of the availabilities in the matching availabilities set
-    for (int i = 0; i < minuteAvailabilities.length; i++) {
-      // gets availability of current minute (considered available if status matches one within set)
-      thisMinuteAvailable = matchingAvailabilities.contains(minuteAvailabilities[i]);
+  private Collection<TimeRange> findTimesBasedOnOptionalAttendeeAvailability(MeetingRequest request, 
+    Availability[] minuteAvailability, int maxOptionalAttendeesUnavailable) {
+
+    ArrayList<TimeRange> availableTimes = new ArrayList<TimeRange>();
+    int start = 0;
+    boolean wasLastMinuteAvailable = minuteAvailability[0].areMandatoryAttendeesAllAvailable 
+        && minuteAvailability[0].numberOfOptionalAttendeesUnavailable <= maxOptionalAttendeesUnavailable;
+  
+    for (int i = 0; i < minuteAvailability.length; i++) {
+    
+      // checks if the current minute is available by checking that mandatory attendees are available
+      // and that the number of optional attendees that can't make it is under the max
+      boolean currentMinuteAvailable = minuteAvailability[i].areMandatoryAttendeesAllAvailable 
+        && minuteAvailability[i].numberOfOptionalAttendeesUnavailable <= maxOptionalAttendeesUnavailable;
 
       if (wasLastMinuteAvailable) {
         // If the previous minute was available, but the current minute is unavailable or if it's
         // the end of the day, this is the end of an available time range. If the time range is longer 
         // than the required duration, it's recorded as an available time range.
-        if (!thisMinuteAvailable  || i == minuteAvailabilities.length - 1) {
+        if (! currentMinuteAvailable || i == minuteAvailability.length - 1) {
           int end = i;
           int duration = end - start;
 
-          if (duration >= requestDuration) {
-            times.add(TimeRange.fromStartEnd(start, end - 1, true)); // add time range (inclusive of start & end) 
-          }
-          
-          // avoid an inconsistency b/c in case availableMinutes[availableMinutes.length - 1] is true,
-          // you shouldn't set this to false (& that would happen without this if)
-          if (!thisMinuteAvailable) {
-            wasLastMinuteAvailable = false;
+          if (duration >= request.getDuration()) {
+            availableTimes.add(TimeRange.fromStartEnd(start, end - 1, true)); // add time range (inclusive of start & end) 
           }
-        }
-      } else if (thisMinuteAvailable) {
+        
+          wasLastMinuteAvailable = false;
+        }      
+      } else {
         // If the last minute was unavailable, but this minute is available, then this is the beginning
         // of a new available time range, so start will be set to this minute and wasLastMinuteAvailable to true.
-
-        start = i;
-        wasLastMinuteAvailable = true;
+        if (currentMinuteAvailable) {
```

Comment:
Nit: this can be `} else if () {` instead of `} else { if () {`

Reply:
Thanks! Will change this.

Reply:
Done